crates/ruff_python_parser/src/parser/expression.rs

```diff
@@ -105,6 +106,99 @@ pub(super) const END_EXPR_SET: TokenSet = TokenSet::new([
 /// Tokens that can appear at the end of a sequence.
 const END_SEQUENCE_SET: TokenSet = END_EXPR_SET.remove(TokenKind::Comma);
 
+/// Represents the expression parsing context.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub(super) struct ExpressionContext(ExpressionContextFlags);
+
+impl Default for ExpressionContext {
+    /// Sets the default flags for the expression context which is to:
+    /// * Consider `in` keyword in a comparison expression
+    /// * Disallow starred and yield expressions
+    fn default() -> Self {
+        ExpressionContext(ExpressionContextFlags::INCLUDE_IN)
+    }
+}
+
+bitflags! {
+    #[derive(Debug, Copy, Clone, PartialEq, Eq)]
+    struct ExpressionContextFlags: u8 {
+        /// This flag is set when the `in` keyword should be included as part of a comparison
+        /// expression. It is to avoid ambiguity in `for ... in ...` statements. This flag is
+        /// set by default.
+        const INCLUDE_IN = 1 << 0;
+
+        /// This flag is set when a starred expression should be allowed. This doesn't affect the
+        /// parsing of a starred expression as it will be parsed nevertheless. But, if it is not
+        /// allowed, an error is reported.
+        const ALLOW_STARRED_EXPRESSION = 1 << 1;
+
+        /// This flag is set when the value of a starred expression should be limited to bitwise OR
+        /// precedence. Matches the `* bitwise_or` grammar rule if set.
+        const BITWISE_OR_PRECEDENCE = 1 << 2;
+
+        /// This flag is set when a yield expression should be allowed. This doesn't affect the
+        /// parsing of a yield expression as it will be parsed nevertheless. But, if it is not
+        /// allowed, an error is reported.
+        const ALLOW_YIELD_EXPRESSION = 1 << 3;
+    }
+}
+
+impl ExpressionContext {
+    /// Returns a new [`ExpressionContext`] which allows starred expression with the given
+    /// precedence.
+    pub(super) fn with_starred_expression_allowed(
+        self,
+        precedence: StarredExpressionPrecedence,
+    ) -> Self {
+        let mut flags = self.0 | ExpressionContextFlags::ALLOW_STARRED_EXPRESSION;
+        if precedence.is_bitwise_or() {
+            flags |= ExpressionContextFlags::BITWISE_OR_PRECEDENCE;
+        }
+        ExpressionContext(flags)
```

Comment:
I think this method is incorrect in the sense that it only adds the `ALLOW_STARRED_EXPRESSION` flags but never removes them. We should either rename the method or ensure to correctly unset the flags (at least the `BITWISE_OR_PRECEDENCE`)

Reply:
Good catch! I'll change it to remove the flag in the `StarredExpressionPrecedence::Conditional` case and use a `match` expression to make it explicit in case any new variant is added in the future.